### DISCUSSION TITLE
CA1862 warning is annoying should the compiler just convert it instead

### DISCUSSION THREAD

**Discussion by AJ1000** [dotnet/csharplang#7711] [Language Ideas]
The CA1862 warning about using string.Equals for case insensitive is annoying, makes the code less readable and flawed as it breaks my EF Core projections (where I am making the database use case insensitivity inside the SQL that EF Core generates).

Since you can detect the case insensitivity in the compiler, can you please convert the code in the IL? That way we have the efficient code without losing the elegance.

This means the code is more elegant to read (as Mads says is his aim in dotnet conf 2023 and elsewhere).

For example I think this is more elegant and readable than:

`if (request.UserEmail.ToLower() == "david.fowler@microsoft.com"))`

than

`if (request.UserEmail.Equals("david.fowler@microsoft.com", StringComparison.CurrentCultureIgnoreCase))`

and have a real dislike for this warning. The code change also breaks my EF Core projections, but I may be doing something wrong there, I am not sure.

Note: The CA!862 warning was introduced in dotnet 8.0 (see https://devblogs.microsoft.com/dotnet/performance-improvements-in-net-8/).

Here is what it says:

CA1862, added in [dotnet/roslyn-analyzers#6662](https://github.com/dotnet/roslyn-analyzers/pull/6662), looks for places where code is performing a case-insensitive comparison (which is fine) but doing so by first lower/uppercasing an input string and then comparing that (which is far from fine). 

It’s much more efficient to just use a StringComparison. [dotnet/runtime#89539](https://github.com/dotnet/runtime/pull/89539) fixes a few such cases.

Thanks.

**Comment by huoyaoyuan** [comment]
CAxxxx analyzers belong to roslyn-analyzers repo. It's not a part of the language.

**Comment by AJ1000** [comment]
You misunderstood. This is not an analyzer issue, as I can disable it. I want the language improved so 

`if (request.UserEmail.ToLower() == "david.fowler@microsoft.com"))`

is handled elegantly and efficiently. This would also mean that the analyzer would no longer be needed.

**Reply by huoyaoyuan** [reply]
This is also not a language issue, since `ToLower` method doesn't belong to the language. It belongs to the runtime.

> is handled elegantly and efficiently.

I'm not sure if an interceptor is able to rewrite it into more efficient form. The language itself disallows any transformations since it has no knowledge about string casing methods.

**Comment by AJ1000** [comment]
Thanks for the information. At least I tried. 

As an end user ToLower() sounds like it is a language feature, so it should be able to be improved and not have the extra memory usage. 

I would have thought the language team would agree that

'if (request.UserEmail.Equals("david.fowler@microsoft.com", StringComparison.CurrentCultureIgnoreCase))'

is not very elegant.

Very frustrating, I will just disable the analyzer and close this issue.
